File: cocos/asset/assets/rendering-sub-mesh.ts
/*
 Copyright (c) 2017-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

import { mapBuffer } from '../../3d/misc/buffer';
import {
    Attribute, Device, InputAssemblerInfo, Buffer, BufferInfo, AttributeName, BufferUsageBit,
    Format, FormatInfos, MemoryUsageBit, PrimitiveMode, DrawInfo,
} from '../../gfx';
import { Vec3, approx, cclegacy, floatToHalf, halfToFloat, pseudoRandomRange } from '../../core';
import { Mesh } from '../../3d/assets/mesh';
import { Root } from '../../root';

/**
 * @en Array views for index buffer
 * @zh 允许存储索引的数组视图。
 */
export type IBArray = Uint8Array | Uint16Array | Uint32Array;

/**
 * @en The interface of geometric information.
 * @zh 几何信息。
 */
export interface IGeometricInfo {
    /**
     * @en Vertex positions.
     * @zh 顶点位置。
     */
    positions: Float32Array;

    /**
     * @en Indices data.
     * @zh 索引数据。
     */
    indices?: IBArray;

    /**
     * @en Whether the geometry is treated as double sided.
     * @zh 是否将图元按双面对待。
     */
    doubleSided?: boolean;

    /**
     * @en The bounding box.
     * @zh 此几何体的轴对齐包围盒。
     */
    boundingBox: { max: Vec3 | Readonly<Vec3>; min: Vec3 | Readonly<Vec3> };
}

/**
 * @en Flat vertex buffer.
 * @zh 扁平化顶点缓冲区。
 */
export interface IFlatBuffer {
    stride: number;
    count: number;
    buffer: Uint8Array;
}

/**
 * @en Sub mesh for rendering which contains all geometry data, it can be used to create [[gfx.InputAssembler]].
 * @zh 包含所有顶点数据的渲染子网格，可以用来创建 [[gfx.InputAssembler]]。
 */
export class RenderingSubMesh {
    /**
     * @en
     * mesh object where this sub mesh locates.
     * @zh
     * 子网格所处的网格模型对象。
     */
    public mesh?: Mesh;

    /**
     * @en
     * sub mesh's index in mesh.
     * @zh
     * 子网格在网格模型中的索引。
     */
    public subMeshIdx?: number;

    private _flatBuffers: IFlatBuffer[] = [];

    private _jointMappedBuffers?: Buffer[];

    private _jointMappedBufferIndices?: number[];

    private _vertexIdChannel?: { stream: number; index: number };

    private _geometricInfo?: IGeometricInfo;

    private _vertexBuffers: Buffer[];

    private _attributes: Attribute[];

    private _indexBuffer: Buffer | null = null;

    private _indirectBuffer: Buffer | null = null;

    private _primitiveMode: PrimitiveMode;

    private _iaInfo: InputAssemblerInfo;

    private _isOwnerOfIndexBuffer = true;

    private _drawInfo?: DrawInfo | null = null;

    private static EMPTY_GEOMETRIC_INFO: IGeometricInfo = {
        positions: new Float32Array(),
        indices: new Uint8Array(),
        boundingBox: { min: Vec3.ZERO, max: Vec3.ZERO },
    };

    /**
     * @en
     * sub mesh's constructor.
     * @zh
     * 子网格构造函数。
     * @param vertexBuffers @en vertex buffers. @zh 顶点缓冲区数组。
     * @param attributes @en vertex attributes. @zh 顶点属性数组。
     * @param primitiveMode @en primitive mode. @zh 图元类型。
     * @param indexBuffer @en index buffer. @zh 索引缓冲区。
     * @param indirectBuffer @en indirect buffer. @zh 间接缓冲区。
     */
    constructor (
        vertexBuffers: Buffer[],
        attributes: Attribute[],
        primitiveMode: PrimitiveMode,
        indexBuffer: Buffer | null = null,
        indirectBuffer: Buffer | null = null,
        isOwnerOfIndexBuffer = true,
    ) {
        this._attributes = attributes;
        this._vertexBuffers = vertexBuffers;
        this._indexBuffer = indexBuffer;
        this._indirectBuffer = indirectBuffer;
        this._primitiveMode = primitiveMode;
        this._iaInfo = new InputAssemblerInfo(attributes, vertexBuffers, indexBuffer, indirectBuffer);
        this._isOwnerOfIndexBuffer = isOwnerOfIndexBuffer;
    }

    /**
     * @en All vertex attributes used by the sub mesh.
     * @zh 所有顶点属性。
     */
    get attributes (): Attribute[] { return this._attributes; }

    /**
     * @en All vertex buffers used by the sub mesh.
     * @zh 使用的所有顶点缓冲区。
     */
    get vertexBuffers (): Buffer[] { return this._vertexBuffers; }

    /**
     * @en Index buffer used by the sub mesh.
     * @zh 使用的索引缓冲区，若未使用则无需指定。
     */
    get indexBuffer (): Buffer | null { return this._indexBuffer; }

    /**
     * @en Indirect buffer used by the sub mesh.
     * @zh 间接绘制缓冲区。
     */
    get indirectBuffer (): Buffer | null { return this._indirectBuffer; }

    /**
     * @en Primitive mode used by the sub mesh.
     * @zh 图元类型。
     */
    get primitiveMode (): PrimitiveMode { return this._primitiveMode; }

    /**
     * @en The geometric info of the sub mesh, used for raycast.
     * @zh （用于射线检测的）几何信息。
     */
    get geometricInfo (): IGeometricInfo {
        if (this._geometricInfo) {
            return this._geometricInfo;
        }
        if (this.mesh === undefined) {
            return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
        }
        if (this.subMeshIdx === undefined) {
            return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
        }
        const { mesh } = this; const index = this.subMeshIdx;
        const pAttri = this.attributes.find((element) => element.name === (AttributeName.ATTR_POSITION as string));

        if (!pAttri) {
            return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
        }

        let positions: Float32Array | undefined;
        switch (pAttri.format) {
        case Format.RG32F:
        case Format.RGB32F:
        {
            positions = mesh.readAttribute(index, AttributeName.ATTR_POSITION) as unknown as Float32Array;
            if (!positions) {
                return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
            }
            break;
        }
        case Format.RGBA32F:
        {
            const data = mesh.readAttribute(index, AttributeName.ATTR_POSITION) as unknown as Float32Array;
            if (!data) {
                return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
            }
            const count = data.length / 4;
            positions = new Float32Array(count * 3);
            for (let i = 0; i < count; ++i) {
                const dstPtr = i * 3;
                const srcPtr = i * 4;
                positions[dstPtr] = data[srcPtr];
                positions[dstPtr + 1] = data[srcPtr + 1];
                positions[dstPtr + 2] = data[srcPtr + 2];
            }
            break;
        }
        case Format.RG16F:
        case Format.RGB16F:
        {
            const data =  mesh.readAttribute(index, AttributeName.ATTR_POSITION) as unknown as Uint16Array;
            if (!data) {
                return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
            }
            positions = new Float32Array(data.length);
            for (let i = 0; i < data.length; ++i) {
                positions[i] = halfToFloat(data[i]);
            }
            break;
        }
        case Format.RGBA16F:
        {
            const data =  mesh.readAttribute(index, AttributeName.ATTR_POSITION) as unknown as Uint16Array;
            if (!data) {
                return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
            }
            const count = data.length / 4;
            positions = new Float32Array(count * 3);
            for (let i = 0; i < count; ++i) {
                const dstPtr = i * 3;
                const srcPtr = i * 4;
                positions[dstPtr] = halfToFloat(data[srcPtr]);
                positions[dstPtr + 1] = halfToFloat(data[srcPtr + 1]);
                positions[dstPtr + 2] = halfToFloat(data[srcPtr + 2]);
            }
            break;
        }
        default:
            return RenderingSubMesh.EMPTY_GEOMETRIC_INFO;
        }

        const indices = mesh.readIndices(index) || undefined;
        const max = new Vec3();
        const min = new Vec3();

        const conut = FormatInfos[pAttri.format].count;
        if (conut === 2) {
            max.set(positions[0], positions[1], 0);
            min.set(positions[0], positions[1], 0);
        } else {
            max.set(positions[0], positions[1], positions[2]);
            min.set(positions[0], positions[1], positions[2]);
        }
        for (let i = 0; i < positions.length; i += conut) {
            if (conut === 2) {
                max.x = positions[i] > max.x ? positions[i] : max.x;
                max.y = positions[i + 1] > max.y ? positions[i + 1] : max.y;
                min.x = positions[i] < min.x ? positions[i] : min.x;
                min.y = positions[i + 1] < min.y ? positions[i + 1] : min.y;
            } else {
                max.x = positions[i] > max.x ? positions[i] : max.x;
                max.y = positions[i + 1] > max.y ? positions[i + 1] : max.y;
                max.z = positions[i + 2] > max.z ? positions[i + 2] : max.z;
                min.x = positions[i] < min.x ? positions[i] : min.x;
                min.y = positions[i + 1] < min.y ? positions[i + 1] : min.y;
                min.z = positions[i + 2] < min.z ? positions[i + 2] : min.z;
            }
        }
        this._geometricInfo = { positions, indices, boundingBox: { max, min } };
        return this._geometricInfo;
    }

    /**
     * @en Invalidate the geometric info of the sub mesh after geometry changed.
     * @zh 网格更新后，设置（用于射线检测的）几何信息为无效，需要重新计算。
     */
    public invalidateGeometricInfo (): void { this._geometricInfo = undefined; }

    /**
     * @en the draw range.
     * @zh 渲染范围。
     */
    set drawInfo (info: DrawInfo | null | undefined) {
        this._drawInfo = info;
    }

    get drawInfo (): DrawInfo | null | undefined {
        return this._drawInfo;
    }

    /**
     * @en Flatted vertex buffers.
     * @zh 扁平化的顶点缓冲区。
     */
    get flatBuffers (): IFlatBuffer[] { return this._flatBuffers; }

    /**
     * @en generate flatted vertex buffers.
     * @zh 生成扁平化的顶点缓冲区。
     */
    public genFlatBuffers (): void {
        if (this._flatBuffers.length || !this.mesh || this.subMeshIdx === undefined) { return; }

        const { mesh } = this;
        let idxCount = 0;
        const prim = mesh.struct.primitives[this.subMeshIdx];
        if (prim.indexView) { idxCount = prim.indexView.count; }
        for (let i = 0; i < prim.vertexBundelIndices.length; i++) {
            const bundleIdx = prim.vertexBundelIndices[i];
            const vertexBundle = mesh.struct.vertexBundles[bundleIdx];
            const vbCount = prim.indexView ? prim.indexView.count : vertexBundle.view.count;
            const vbStride = vertexBundle.view.stride;
            const vbSize = vbStride * vbCount;
            const view = new Uint8Array(mesh.data.buffer, vertexBundle.view.offset, vertexBundle.view.length);
            const sharedView = new Uint8Array(prim.indexView ? vbSize : vertexBundle.view.length);

            if (!prim.indexView) {
                sharedView.set(mesh.data.subarray(vertexBundle.view.offset, vertexBundle.view.offset + vertexBundle.view.length));
                this._flatBuffers.push({ stride: vbStride, count: vbCount, buffer: sharedView });
                continue;
            }

            const ibView = mesh.readIndices(this.subMeshIdx)!;
            // transform to flat buffer
            for (let n = 0; n < idxCount; ++n) {
                const idx = ibView[n];
                const offset = n * vbStride;
                const srcOffset = idx * vbStride;
                for (let m = 0; m < vbStride; ++m) {
                    sharedView[offset + m] = view[srcOffset + m];
                }
            }
            this._flatBuffers.push({ stride: vbStride, count: vbCount, buffer: sharedView });
        }
    }

    /**
     * @en The vertex buffer for joint after mapping.
     * @zh 骨骼索引按映射表处理后的顶点缓冲。
     */
    get jointMappedBuffers (): Buffer[] {
        if (this._jointMappedBuffers) { return this._jointMappedBuffers; }
        const buffers: Buffer[] = this._jointMappedBuffers = [];
        const indices: number[] = this._jointMappedBufferIndices = [];
        if (!this.mesh || this.subMeshIdx === undefined) { return this._jointMappedBuffers = this.vertexBuffers; }
        const { struct } = this.mesh;
        const prim = struct.primitives[this.subMeshIdx];
        if (!struct.jointMaps || prim.jointMapIndex === undefined || !struct.jointMaps[prim.jointMapIndex]) {
            return this._jointMappedBuffers = this.vertexBuffers;
        }
        let jointFormat: Format;
        let jointOffset: number;
        const { device } = cclegacy.director.root as Root;
        for (let i = 0; i < prim.vertexBundelIndices.length; i++) {
            const bundle = struct.vertexBundles[prim.vertexBundelIndices[i]];
            jointOffset = 0;
            jointFormat = Format.UNKNOWN;
            for (let j = 0; j < bundle.attributes.length; j++) {
                const attr = bundle.attributes[j];
                if (attr.name === (AttributeName.ATTR_JOINTS as string)) {
                    jointFormat = attr.format;
                    break;
                }
                jointOffset += FormatInfos[attr.format].size;
            }
            if (jointFormat) {
                const data = new Uint8Array(this.mesh.data.buffer, bundle.view.offset, bundle.view.length);
                const dataView = new DataView(data.slice().buffer);
                const idxMap = struct.jointMaps[prim.jointMapIndex];
                mapBuffer(
                    dataView,
                    (cur): number => idxMap.indexOf(cur),
                    jointFormat,
                    jointOffset,
                    bundle.view.length,
                    bundle.view.stride,
                    dataView,
                );
                const buffer = device.createBuffer(new BufferInfo(
                    BufferUsageBit.VERTEX | BufferUsageBit.TRANSFER_DST,
                    MemoryUsageBit.DEVICE,
                    bundle.view.length,
                    bundle.view.stride,
                ));
                buffer.update(dataView.buffer); buffers.push(buffer); indices.push(i);
            } else {
                buffers.push(this.vertexBuffers[prim.vertexBundelIndices[i]]);
            }
        }
        if (this._vertexIdChannel) {
            buffers.push(this._allocVertexIdBuffer(device));
        }
        return buffers;
    }

    /**
     * @en The input assembler info.
     * @zh 输入汇集器信息。
     */
    get iaInfo (): InputAssemblerInfo { return this._iaInfo; }

    /**
     * @en Destroys sub mesh.
     * @zh 销毁子网格。
     */
    public destroy (): void {
        for (let i = 0; i < this.vertexBuffers.length; i++) {
            this.vertexBuffers[i].destroy();
        }
        this.vertexBuffers.length = 0;
        if (this._indexBuffer) {
            if (this._isOwnerOfIndexBuffer) {
                this._indexBuffer.destroy();
            }
            this._indexBuffer = null;
        }
        if (this._jointMappedBuffers && this._jointMappedBufferIndices) {
            for (let i = 0; i < this._jointMappedBufferIndices.length; i++) {
                this._jointMappedBuffers[this._jointMappedBufferIndices[i]].destroy();
            }
            this._jointMappedBuffers = undefined;
            this._jointMappedBufferIndices = undefined;
        }
        if (this._indirectBuffer) {
            this._indirectBuffer.destroy();
            this._indirectBuffer = null;
        }
    }

    /**
     * @en Adds a vertex attribute input called 'a_vertexId' into this sub-mesh.
     * This is useful if you want to simulate `gl_VertexId` in WebGL context prior to 2.0.
     * Once you call this function, the vertex attribute is permanently added.
     * Subsequent calls to this function take no effect.
     * @zh 添加一个 'a_vertexId' 顶点属性， 用于在 WebGL 2.0 之前的平台模拟 `gl_VertexId`，
     * 一旦你调用此函数， 顶点属性永久被添加， 后续调用无效果。
     * @param device @en Device used to create related rendering resources @zh 用于创建相关渲染资源的设备对象
     */
    public enableVertexIdChannel (device: Device): void {
        if (this._vertexIdChannel) {
            return;
        }

        const streamIndex = this.vertexBuffers.length;
        const attributeIndex = this.attributes.length;

        const vertexIdBuffer = this._allocVertexIdBuffer(device);
        this._vertexBuffers.push(vertexIdBuffer);
        this._attributes.push(new Attribute('a_vertexId', Format.R32F, false, streamIndex));

        this._iaInfo.attributes = this._attributes;
        this._iaInfo.vertexBuffers = this._vertexBuffers;

        this._vertexIdChannel = {
            stream: streamIndex,
            index: attributeIndex,
        };
    }

    private _allocVertexIdBuffer (device: Device): Buffer {
        const vertexCount = (this.vertexBuffers.length === 0 || this.vertexBuffers[0].stride === 0)
            ? 0
            // TODO: This depends on how stride of a vertex buffer is defined; Consider padding problem.
            : this.vertexBuffers[0].size / this.vertexBuffers[0].stride;
        const vertexIds = new Float32Array(vertexCount);
        for (let iVertex = 0; iVertex < vertexCount; ++iVertex) {
            // `+0.5` because on some platforms, the "fetched integer" may have small error.
            // For example `26` may yield `25.99999`, which is convert to `25` instead of `26` using `int()`.
            vertexIds[iVertex] = iVertex + 0.5;
        }

        const vertexIdBuffer = device.createBuffer(new BufferInfo(
            BufferUsageBit.VERTEX | BufferUsageBit.TRANSFER_DST,
            MemoryUsageBit.DEVICE,
            vertexIds.byteLength,
            vertexIds.BYTES_PER_ELEMENT,
        ));
        vertexIdBuffer.update(vertexIds);

        return vertexIdBuffer;
    }
}

File: native/cocos/core/assets/RenderingSubMesh.cpp
/****************************************************************************
 Copyright (c) 2021-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "core/assets/RenderingSubMesh.h"
#include <cstdint>
#include "3d/assets/Mesh.h"
#include "3d/misc/Buffer.h"
#include "core/DataView.h"
#include "core/TypedArray.h"
#include "math/Utils.h"
#include "math/Vec3.h"
#include "renderer/gfx-base/GFXBuffer.h"
#include "renderer/gfx-base/GFXDevice.h"

namespace cc {

RenderingSubMesh::RenderingSubMesh(const gfx::BufferList &vertexBuffers,
                                   const gfx::AttributeList &attributes,
                                   gfx::PrimitiveMode primitiveMode)
: RenderingSubMesh(vertexBuffers, attributes, primitiveMode, nullptr, nullptr, true) {
}

RenderingSubMesh::RenderingSubMesh(const gfx::BufferList &vertexBuffers,
                                   const gfx::AttributeList &attributes,
                                   gfx::PrimitiveMode primitiveMode,
                                   gfx::Buffer *indexBuffer)
: RenderingSubMesh(vertexBuffers, attributes, primitiveMode, indexBuffer, nullptr, true) {
}

RenderingSubMesh::RenderingSubMesh(const gfx::BufferList &vertexBuffers,
                                   const gfx::AttributeList &attributes,
                                   gfx::PrimitiveMode primitiveMode,
                                   gfx::Buffer *indexBuffer,
                                   gfx::Buffer *indirectBuffer)
: RenderingSubMesh(vertexBuffers, attributes, primitiveMode, indexBuffer, indirectBuffer, true) {
}

RenderingSubMesh::RenderingSubMesh(const gfx::BufferList &vertexBuffers,
                                   const gfx::AttributeList &attributes,
                                   gfx::PrimitiveMode primitiveMode,
                                   gfx::Buffer *indexBuffer,
                                   gfx::Buffer *indirectBuffer,
                                   bool isOwnerOfIndexBuffer)
: _vertexBuffers(vertexBuffers),
  _attributes(attributes),
  _primitiveMode(primitiveMode),
  _indexBuffer(indexBuffer),
  _indirectBuffer(indirectBuffer),
  _isOwnerOfIndexBuffer(isOwnerOfIndexBuffer) {
    _iaInfo.attributes = attributes;
    _iaInfo.vertexBuffers = vertexBuffers;
    _iaInfo.indexBuffer = indexBuffer;
    _iaInfo.indirectBuffer = indirectBuffer;
}

RenderingSubMesh::~RenderingSubMesh() {
    destroy();
}

const IGeometricInfo &RenderingSubMesh::getGeometricInfo() {
    if (_geometricInfo.has_value()) {
        return _geometricInfo.value();
    }
    // NOLINTNEXTLINE
    static const IGeometricInfo EMPTY_GEOMETRIC_INFO;
    if (_mesh == nullptr) {
        return EMPTY_GEOMETRIC_INFO;
    }

    if (!_subMeshIdx.has_value()) {
        return EMPTY_GEOMETRIC_INFO;
    }

    auto iter = std::find_if(_attributes.cbegin(), _attributes.cend(), [](const gfx::Attribute &element) -> bool {
        return element.name == gfx::ATTR_NAME_POSITION;
    });
    if (iter == _attributes.end()) {
        return EMPTY_GEOMETRIC_INFO;
    }

    const auto &attri = *iter;
    const uint32_t count = gfx::GFX_FORMAT_INFOS[static_cast<uint32_t>(attri.format)].count;

    auto index = static_cast<index_t>(_subMeshIdx.value());
    const auto &positionsVar = _mesh->readAttribute(index, gfx::ATTR_NAME_POSITION);

    Float32Array const *pPositions = nullptr;
    switch (attri.format) {
        case gfx::Format::RG32F:
        case gfx::Format::RGB32F: {
            pPositions = ccstd::get_if<Float32Array>(&positionsVar);
            if (pPositions == nullptr) {
                return EMPTY_GEOMETRIC_INFO;
            }
            break;
        }
        case gfx::Format::RGBA32F: {
            const auto *data = ccstd::get_if<Float32Array>(&positionsVar);
            if (data == nullptr) {
                return EMPTY_GEOMETRIC_INFO;
            }
            const auto count = data->length() / 4;
            auto *pos = ccnew Float32Array(count * 3);
            for (uint32_t i = 0; i < count; i++) {
                const auto dstPtr = i * 3;
                const auto srcPtr = i * 4;
                (*pos)[dstPtr] = (*data)[srcPtr];
                (*pos)[dstPtr + 1] = (*data)[srcPtr + 1];
                (*pos)[dstPtr + 2] = (*data)[srcPtr + 2];
            }
            pPositions = pos;
            break;
        }
        case gfx::Format::RG16F:
        case gfx::Format::RGB16F: {
            const auto *data = ccstd::get_if<Uint16Array>(&positionsVar);
            if (data == nullptr) {
                return EMPTY_GEOMETRIC_INFO;
            }
            auto *pos = ccnew Float32Array(data->length());
            for (uint32_t i = 0; i < data->length(); ++i) {
                (*pos)[i] = mathutils::halfToFloat((*data)[i]);
            }
            pPositions = pos;
            break;
        }
        case gfx::Format::RGBA16F: {
            const auto *data = ccstd::get_if<Uint16Array>(&positionsVar);
            if (data == nullptr) {
                return EMPTY_GEOMETRIC_INFO;
            }
            const auto count = data->length() / 4;
            auto *pos = ccnew Float32Array(count * 3);
            for (uint32_t i = 0; i < count; i++) {
                const auto dstPtr = i * 3;
                const auto srcPtr = i * 4;
                (*pos)[dstPtr] = mathutils::halfToFloat((*data)[srcPtr]);
                (*pos)[dstPtr + 1] = mathutils::halfToFloat((*data)[srcPtr + 1]);
                (*pos)[dstPtr + 2] = mathutils::halfToFloat((*data)[srcPtr + 2]);
            }
            pPositions = pos;
            break;
        }
        default:
            return EMPTY_GEOMETRIC_INFO;
    };

    const auto &positions = *pPositions;
    const auto &indicesVar = _mesh->readIndices(index);

    Vec3 max;
    Vec3 min;

    if (count == 2) {
        max.set(positions[0], positions[1], 0);
        min.set(positions[0], positions[1], 0);
    } else {
        max.set(positions[0], positions[1], positions[2]);
        min.set(positions[0], positions[1], positions[2]);
    }

    for (int i = 0; i < positions.length(); i += static_cast<int>(count)) {
        if (count == 2) {
            max.x = positions[i] > max.x ? positions[i] : max.x;
            max.y = positions[i + 1] > max.y ? positions[i + 1] : max.y;
            min.x = positions[i] < min.x ? positions[i] : min.x;
            min.y = positions[i + 1] < min.y ? positions[i + 1] : min.y;
        } else {
            max.x = positions[i] > max.x ? positions[i] : max.x;
            max.y = positions[i + 1] > max.y ? positions[i + 1] : max.y;
            max.z = positions[i + 2] > max.z ? positions[i + 2] : max.z;
            min.x = positions[i] < min.x ? positions[i] : min.x;
            min.y = positions[i + 1] < min.y ? positions[i + 1] : min.y;
            min.z = positions[i + 2] < min.z ? positions[i + 2] : min.z;
        }
    }

    IGeometricInfo info;
    info.positions = positions;
    info.indices = indicesVar;
    info.boundingBox.max = max;
    info.boundingBox.min = min;

    _geometricInfo = info;
    return _geometricInfo.value();
}

void RenderingSubMesh::genFlatBuffers() {
    if (!_flatBuffers.empty() || _mesh == nullptr || !_subMeshIdx.has_value()) {
        return;
    }

    uint32_t idxCount = 0;
    const auto &prim = _mesh->getStruct().primitives[_subMeshIdx.value()];
    if (prim.indexView.has_value()) {
        idxCount = prim.indexView.value().count;
    }
    for (size_t i = 0; i < prim.vertexBundelIndices.size(); i++) {
        const uint32_t bundleIdx = prim.vertexBundelIndices[i];
        const Mesh::IVertexBundle &vertexBundle = _mesh->getStruct().vertexBundles[bundleIdx];
        const uint32_t vbCount = prim.indexView.has_value() ? prim.indexView.value().count : vertexBundle.view.count;
        const uint32_t vbStride = vertexBundle.view.stride;
        const uint32_t vbSize = vbStride * vbCount;
        Uint8Array view(_mesh->getData().buffer(), vertexBundle.view.offset, vertexBundle.view.length);
        Uint8Array sharedView(prim.indexView.has_value() ? vbSize : vertexBundle.view.length);

        if (!prim.indexView.has_value()) {
            sharedView.set(_mesh->getData().subarray(vertexBundle.view.offset, vertexBundle.view.offset + vertexBundle.view.length));
            _flatBuffers.emplace_back(IFlatBuffer{vbStride, vbCount, sharedView});
            continue;
        }

        IBArray ibView = _mesh->readIndices(static_cast<int>(_subMeshIdx.value()));
        // transform to flat buffer
        for (uint32_t n = 0; n < idxCount; ++n) {
            auto idx = getIBArrayValue<int32_t>(ibView, static_cast<int>(n));
            uint32_t offset = n * vbStride;
            uint32_t srcOffset = idx * vbStride;
            for (uint32_t m = 0; m < vbStride; ++m) {
                sharedView[static_cast<int>(offset + m)] = view[static_cast<int>(srcOffset + m)];
            }
        }
        _flatBuffers.emplace_back(IFlatBuffer{vbStride, vbCount, std::move(sharedView)});
    }
}

void RenderingSubMesh::enableVertexIdChannel(gfx::Device *device) {
    if (_vertexIdChannel.has_value()) {
        return;
    }

    const auto streamIndex = static_cast<uint32_t>(_vertexBuffers.size());
    const auto attributeIndex = static_cast<uint32_t>(_attributes.size());

    gfx::Buffer *vertexIdBuffer = allocVertexIdBuffer(device);
    _vertexBuffers.pushBack(vertexIdBuffer);
    _attributes.push_back({"a_vertexId", gfx::Format::R32F, false, streamIndex, false, 0});

    _iaInfo.attributes = _attributes;
    _iaInfo.vertexBuffers = _vertexBuffers.get();

    _vertexIdChannel = VertexIdChannel{
        streamIndex,
        attributeIndex,
    };
}

bool RenderingSubMesh::destroy() {
    _vertexBuffers.clear();
    _indexBuffer = nullptr;
    _indirectBuffer = nullptr;

    if (!_jointMappedBuffers.empty() && !_jointMappedBufferIndices.empty()) {
        for (uint32_t index : _jointMappedBufferIndices) {
            _jointMappedBuffers.at(index)->destroy();
        }
        _jointMappedBuffers.clear();
        _jointMappedBufferIndices.clear();
    }
    return true;
}

const gfx::BufferList &RenderingSubMesh::getJointMappedBuffers() {
    if (!_jointMappedBuffers.empty()) {
        return _jointMappedBuffers.get();
    }

    auto &buffers = _jointMappedBuffers;
    auto &indices = _jointMappedBufferIndices;

    if (!_mesh || !_subMeshIdx.has_value()) {
        _jointMappedBuffers = _vertexBuffers;
        return _jointMappedBuffers.get();
    }

    const auto &structInfo = _mesh->getStruct();
    const auto &prim = structInfo.primitives[_subMeshIdx.value()];
    if (!structInfo.jointMaps.has_value() || !prim.jointMapIndex.has_value() || structInfo.jointMaps.value()[prim.jointMapIndex.value()].empty()) {
        _jointMappedBuffers = _vertexBuffers;
        return _jointMappedBuffers.get();
    }
    gfx::Format jointFormat = gfx::Format::UNKNOWN;
    int32_t jointOffset = 0;
    gfx::Device *device = gfx::Device::getInstance();
    for (size_t i = 0; i < prim.vertexBundelIndices.size(); i++) {
        const auto &bundle = structInfo.vertexBundles[prim.vertexBundelIndices[i]];
        jointOffset = 0;
        jointFormat = gfx::Format::UNKNOWN;
        for (const auto &attr : bundle.attributes) {
            if (attr.name == gfx::ATTR_NAME_JOINTS) {
                jointFormat = attr.format;
                break;
            }
            jointOffset += static_cast<int32_t>(gfx::GFX_FORMAT_INFOS[static_cast<int32_t>(attr.format)].size);
        }
        if (jointFormat != gfx::Format::UNKNOWN) {
            Uint8Array data{_mesh->getData().buffer(), bundle.view.offset, bundle.view.length};
            DataView dataView(data.slice().buffer());
            const auto &idxMap = structInfo.jointMaps.value()[prim.jointMapIndex.value()];

            mapBuffer(
                dataView, [&](const DataVariant &cur, uint32_t /*idx*/, const DataView & /*view*/) -> DataVariant {
                    if (ccstd::holds_alternative<int32_t>(cur)) {
                        auto iter = std::find(idxMap.begin(), idxMap.end(), ccstd::get<int32_t>(cur));
                        if (iter != idxMap.end()) {
                            return static_cast<int32_t>(iter - idxMap.begin());
                        }
                    }
                    CC_ABORT();
                    return -1;
                },
                jointFormat, jointOffset, bundle.view.length, bundle.view.stride, &dataView);

            auto *buffer = device->createBuffer(gfx::BufferInfo{
                gfx::BufferUsageBit::VERTEX | gfx::BufferUsageBit::TRANSFER_DST,
                gfx::MemoryUsageBit::DEVICE,
                bundle.view.length,
                bundle.view.stride});

            buffer->update(dataView.buffer()->getData());
            buffers.pushBack(buffer);
            indices.emplace_back(i);
        } else {
            buffers.pushBack(_vertexBuffers.at(prim.vertexBundelIndices[i]));
        }
    }

    if (_vertexIdChannel) {
        buffers.pushBack(allocVertexIdBuffer(device));
    }
    return buffers.get();
}

gfx::Buffer *RenderingSubMesh::allocVertexIdBuffer(gfx::Device *device) {
    const uint32_t vertexCount = (_vertexBuffers.empty() || _vertexBuffers.at(0)->getStride() == 0)
                                     ? 0
                                     // TODO(minggo): This depends on how stride of a vertex buffer is defined; Consider padding problem.
                                     : _vertexBuffers.at(0)->getSize() / _vertexBuffers.at(0)->getStride();

    ccstd::vector<float> vertexIds(vertexCount);
    for (int iVertex = 0; iVertex < vertexCount; ++iVertex) {
        // `+0.5` because on some platforms, the "fetched integer" may have small error.
        // For example `26` may yield `25.99999`, which is convert to `25` instead of `26` using `int()`.
        vertexIds[iVertex] = static_cast<float>(iVertex) + 0.5F;
    }

    uint32_t vertexIdxByteLength = sizeof(float) * vertexCount;
    gfx::Buffer *vertexIdBuffer = device->createBuffer({
        gfx::BufferUsageBit::VERTEX | gfx::BufferUsageBit::TRANSFER_DST,
        gfx::MemoryUsageBit::DEVICE,
        vertexIdxByteLength,
        sizeof(float),
    });
    vertexIdBuffer->update(vertexIds.data(), vertexIdxByteLength);

    return vertexIdBuffer;
}

} // namespace cc
